README.md
# NAME

App::Jiffy - A minimalist time tracking app focused on precision and effortlessness.

# SYNOPSIS

    use App::Jiffy;

    # cmd line tool
    jiffy Solving world hunger
    jiffy Cleaning the plasma manifolds
    jiffy current # Returns the elapsed time for the current task

    # Run server
    jiffyd
    curl -d "title=Meeting with Client X" http://localhost:3000/timeentry

# DESCRIPTION

App::Jiffy's philosophy is that you should have to do as little as possible to track your time. Instead you should focus on working. App::Jiffy also focuses on precision. Many times time tracking results in globbing activities together masking the fact that your 5 hours of work on project "X" was actually 3 hours of work with interruptions from your coworker asking about project "Y".

In order to be precise with as little effort as possible, App::Jiffy will be available via a myriad of mediums and devices but will have a central server to combine all the information. Plans currently include the following applications:

- Command line tool
- Web app [App::Jiffyd](https://metacpan.org/pod/App::Jiffyd)
- iPhone app ( potentially )

# INSTALLATION

    curl -L https://cpanmin.us | perl - git://github.com/lejeunerenard/jiffy

# METHODS

The following are methods available on the `App::Jiffy` object.

## add\_entry

`add_entry` will create a new TimeEntry with the current time as the entry's start\_time.

## current\_time

`current_time` will print out the elapsed time for the current task (AKA the time since the last entry was created).

## time\_sheet

`time_sheet` will print out a time sheet including the time spent for each `TimeEntry`.

## search( `$query_text`, `$days` )

The `search` subcommand will look for the given `$query_text` in the past `$days` number of days. It will treat the `$query_text` argument as a regex.

## run

`run` will start an instance of the Jiffy app.

# AUTHOR

Sean Zellmer <sean@lejeunerenard.com>

# COPYRIGHT

Copyright 2015- Sean Zellmer

# LICENSE

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.


Changes
Revision history for App-Jiffy

{{$NEXT}}

0.09  2020-03-31 16:46:10 CDT
        - Prevent warning when entering blank entry
        - Set binmode for output to support utf8 strings

0.08  2018-10-31 19:32:13 PDT
        - Set timezone to `local` when displaying today's `timesheet`

0.07  2018-04-06 17:47:15 CDT
        - Upgrade MonogoDB dependency to support v1

0.06  2018-04-06 16:05:57 CDT
        - Fix tests to skip if no MongoDB is present

0.05  2017-12-29 13:10:23 CST
        - Use local timezone in `search()`

0.04  2017-12-29 11:45:04 CST
        - Added rounding to durations
        - Added JSON export

0.03  2016-02-22 14:45:04 CST
        - Added search and timesheet subcommands
        - Fixed add_entry timezone bug
        - Moved duration calculation into TimeEntry
        - Added entry's title to current subcommand output
        - Added tests

0.02  2015-04-06 14:48:30 CDT
        - Added minimalist HTTP server ( AKA `jiffyd` )
        - Added docs
        - Added `current` subcommand to `App::Jiffy`

0.01  2015-04-03 01:14:57 CDT
        - Initial release
        - Insert and Update TimeEntries


t/lib/MongoDBTest.pm
package MongoDBTest;

use strict;
use warnings;

use Exporter 'import';
use Test::More;
use MongoDB;

our @EXPORT_OK = qw(
  test_db_or_skip
);

sub test_db_or_skip {
  my $cfg = shift;
  eval {
    my $client = MongoDB::MongoClient->new;
    my $db     = $client->get_database($cfg->{db});
  };

  if ($@) {
    ( my $err = $@ ) =~ s/\n//g;
    if ( $err =~ /couldn't connect|connection refused/i ) {
      $err = "no mongod on localhost:27017";
    }

    plan skip_all => "$err";
  }
}



t/lib/CreateTimeEntries.pm
package CreateTimeEntries;

use strict;
use warnings;

use App::Jiffy::TimeEntry;

use parent qw( Exporter );

our @EXPORT = qw/generate/;

sub generate {
  my $cfg      = shift;
  my $entries = shift;

  my $now = DateTime->now(time_zone => 'local');

  for my $i (0 .. $#$entries) {
    my $entry = $entries->[$i];
    my $start_time;

    if ( ref $entry->{start_time} eq 'CODE' ) {
      $start_time = $entry->{start_time}->();
    } elsif ( ref $entry->{start_time} eq 'HASH' ) {
      $start_time = $now->clone->subtract( %{$entry->{start_time}} );
    }

    # Default
    $start_time //= $now->clone->subtract(hours => ( scalar( @$entries ) - $i ));

    App::Jiffy::TimeEntry->new(
      title      => $entry->{title} // 'Beep Boop',
      start_time => $start_time,
      cfg        => $cfg,
    )->save;
  }
}

1;


lib/App/Jiffy/TimeEntry.pm
package App::Jiffy::TimeEntry;

use strict;
use warnings;

use Scalar::Util qw( blessed );
use MongoDB;
use MongoDB::OID;

use Moo;

my $collection = 'timeEntry';

has id => (
  is  => 'rw',
  isa => sub {
    die 'id must be a MongoDB::OID'
      unless blessed $_[0] and $_[0]->isa('MongoDB::OID');
  },
);

has start_time => (
  is  => 'rw',
  isa => sub {
    die 'start_time is not a DateTime object'
      unless blessed $_[0] and $_[0]->isa('DateTime');
  },
  default => sub {
    DateTime->now();
  },
);

has title => (
  is       => 'rw',
  required => 1,
);

has _duration => (
  is       => 'rw',
);

has cfg => ( is => 'ro', );

=head2 db

C<db> get the db handler from MongoDB

=cut

sub db {
  my $self   = shift;
  my $client = MongoDB::MongoClient->new;
  return $client->get_database( ( $self->cfg ) ? $self->cfg->{db} : 'jiffy' );
}

=head2 save

C<save> will commit the current state of the C<TimeEntry> to the database.

=cut

sub save {
  my $self     = shift;
  my $_id      = $self->id;
  my $document = {
    start_time => $self->start_time,
    title      => $self->title,
  };
  if ($_id) {

    # Update the existing record
    $self->db->get_collection($collection)
      ->update( { _id => $self->id }, { '$set' => $document } );
  } else {

    # Insert new record
    $_id = $self->db->get_collection($collection)->insert($document);

    # Update id
    $self->id($_id);
  }
}

=head2 duration

C<duration> returns the time between this entry and the next.

=cut

sub duration {
  my $self = shift;
  my $set = shift;

  if ($set) {
    $self->_duration($set);
  }

  if ($self->_duration) {
    return $self->_duration;
  }

  my @entry_after = App::Jiffy::TimeEntry::search(
    $self->cfg,
    query => {
      start_time => {
        '$gt' => $self->start_time,
      },
    },
    sort => {
      start_time => 1,
    },
    limit => 1,
  );

  if (@entry_after) {
    return $entry_after[0]->start_time->subtract_datetime( $self->start_time );
  } else {
    return DateTime->now->subtract_datetime( $self->start_time );
  }
}

=head2 find

C<find> will return a single document. The query will use the provided C<_id>.

=cut

sub find {
  my $cfg = shift;
  my $_id = shift;

  my $client = MongoDB::MongoClient->new;
  my $entry  = $client->get_database( $cfg->{db} )->get_collection($collection)
    ->find_one( { _id => $_id } );
  return unless $entry;
  return App::Jiffy::TimeEntry->new(
    id         => $entry->{_id},
    title      => $entry->{title},
    start_time => $entry->{start_time},
    cfg        => $cfg,
  );
}

=head2 search C<$config> C<%options>...

C<search> will return an array of TimeEntry that fit the given C<%options>.

C<%options> can include the following:

=over

=item C<query>

This is the query syntax available via L<MongoDB::Collection>.

=item C<sort>

If present, the results will be sorted via the hashref given by this option.

=item C<limit>

If present, the results will limited to the number provided.

=back

=cut

sub search {
  my $cfg     = shift;
  my %options = @_;
  my $query   = $options{query};
  my $sort    = $options{sort};
  my $limit   = $options{limit};

  my $client  = MongoDB::MongoClient->new;
  my $entries = $client->get_database( $cfg->{db} )->get_collection($collection)
    ->find($query);

  if ($sort) {
    $entries = $entries->sort($sort);
  }

  if ($limit) {
    $entries = $entries->limit($limit);
  }

  # Return undef if nothing was found
  return unless $entries;

# @TODO create a subclass of the MongoDB cursor that allows chaining of results like MongoDB
  map {
    App::Jiffy::TimeEntry->new(
      id         => $_->{_id},
      title      => $_->{title},
      start_time => $_->{start_time},
      cfg        => $cfg,
      )
  } $entries->all;
}

=head2 last_entry

C<last_entry> will return the last TimeEntry in the db.

=cut

sub last_entry {
  my $cfg = shift;

  my $client = MongoDB::MongoClient->new;
  my $entry  = $client->get_database( $cfg->{db} )->get_collection($collection)
    ->find->sort( { start_time => -1 } )->limit(1)->next;
  return unless $entry;
  return App::Jiffy::TimeEntry->new(
    id         => $entry->{_id},
    title      => $entry->{title},
    start_time => $entry->{start_time},
    cfg        => $cfg,
  );
}

=head2 TO_JSON

C<TO_JSON> will return the entry as a JSON convertible hash.

=cut

sub TO_JSON {
  my $self = shift;

  {
    start_time => $self->start_time->iso8601,
    title => $self->title,
    duration => {
      $self->duration->deltas
    },
  }
}

1;


lib/App/Jiffy/View/Timesheet.pm
package App::Jiffy::View::Timesheet;

use strict;
use warnings;

use App::Jiffy::TimeEntry;
use DateTime;

sub render {
  my $entries = shift;
  my $options = shift;
  my $from = $options->{from};

  # Header
  if ($from) {
    print "The past " . $from . " days' timesheet:\n\n";
  } else {
    print "Today's timesheet:\n\n";
  }

  my $current_day = $entries->[0]->start_time->clone->truncate( to => 'day' );
  if ($from) {
    print "Date: " . $current_day->mdy('/') . "\n";
  }

  foreach my $entry (@$entries) {

    my $start_time = $entry->start_time->clone;

    if (
      DateTime->compare( $current_day, $start_time->truncate( to => 'day' ) )
      == -1 )
    {
      $current_day = $start_time->truncate( to => 'day' );
      print "\nDate: " . $current_day->mdy('/') . "\n";
    }

    # Get the deltas
    my $duration = $entry->duration;

    if ( $options->{round} ) {
      # Round seconds
      my $seconds = $duration->seconds;
      if ( $seconds >= 30 ) {
        $duration->add( minutes => 1 );
      }
      $duration->subtract( seconds => $seconds );

      # Round minutes
      my $minutes = $duration->minutes;
      if ( $minutes % 15 >= 15 / 2 ) {
        $duration->add( minutes => 15 - ($minutes % 15) );
      } else {
        $duration->subtract( minutes => $minutes % 15 );
      }
    }

    my %deltas = $duration->deltas;

    foreach my $unit ( sort keys %deltas ) {
      next unless $deltas{$unit};
      print $deltas{$unit} . " " . $unit . " ";
    }

    # Print entry
    if ( $options->{verbose} ) {
      my ( $clock_time ) = $entry->start_time->hms =~ /(.*):.*$/;
      print "\t " .
      # Time
        $clock_time .
      # Title
        "\t" . $entry->title . "\n";
    } else {
      print "\t " . $entry->title . "\n";
    }
  }
}

1;


lib/App/Jiffy.pm
package App::Jiffy;

use strict;
use warnings;

use utf8;
use open ':std', ':encoding(utf8)';

use 5.008_005;
our $VERSION = '0.09';

use App::Jiffy::TimeEntry;
use App::Jiffy::View::Timesheet;
use App::Jiffy::Util::Duration qw/round/;

use YAML::Any qw( LoadFile );
use JSON::MaybeXS 'JSON';
use Scalar::Util qw( looks_like_number );

use Getopt::Long;
Getopt::Long::Configure("pass_through");

use Moo;

has cfg => (
  is      => 'ro',
  default => sub {
    LoadFile( $ENV{HOME} . '/.jiffy.yml' ) || {};
  },
);

has terminator_regex => (
  is  => 'ro',
  isa => sub {
    die 'terminator_regex must be a regex ref' unless ref $_[0] eq 'Regexp';
  },
  default => sub {
    qr/^end$|
    ^done$|
    ^eod$|
    ^finished$|
    ^\\\(^\s*\.^\s*\)\/$| # This is a smily face with hands raised
    ^✓$|
    ^x$/x;
  },
);

sub remove_terminators {
  my $self = shift;
  return (
    title => {
      '$not' => $self->terminator_regex,
    } );
}

sub add_entry {
  my $self    = shift;
  my $options = shift;
  my $title;
  if ( ref $options ne 'HASH' ) {
    $title = $options;
    undef $options;
  } else {
    $title = shift;
  }

  my $start_time;

  my $LocalTZ = DateTime::TimeZone->new( name => 'local' );    # For caching
  my $now = DateTime->now( time_zone => $LocalTZ );

  if ( $options->{time} ) {
    require DateTime::Format::Strptime;

    # @TODO Figure out something more flexible and powerful to get time

    # First try H:M:S
    my $strp = DateTime::Format::Strptime->new(
      pattern   => '%T',
      time_zone => $LocalTZ,
    );
    $start_time = $strp->parse_datetime( $options->{time} );

    # If no time found try just H:M
    if ( not $start_time ) {
      my $strp = DateTime::Format::Strptime->new(
        pattern   => '%R',
        time_zone => $LocalTZ,
      );
      $start_time = $strp->parse_datetime( $options->{time} );
    }

    # Make sure the date part of the datetime is not set to the
    # beginning of time.
    if ( $start_time and $start_time->year < $now->year ) {
      $start_time->set(
        day   => $now->day,
        month => $now->month,
        year  => $now->year,
      );
    }
  }

  # Create and save Entry
  App::Jiffy::TimeEntry->new(
    title      => $title,
    start_time => $start_time // $now,
    cfg        => $self->cfg,
  )->save;
}

sub current_time {
  my $self = shift;

  my $latest   = App::Jiffy::TimeEntry::last_entry( $self->cfg );
  my $duration = $latest->duration;

  print '"' . $latest->title . '" has been running for';

  my %deltas = $duration->deltas;
  foreach my $unit ( keys %deltas ) {
    next unless $deltas{$unit};
    print " " . $deltas{$unit} . " " . $unit;
  }
  print ".\n";
}

sub time_sheet {
  my $self    = shift;
  my $options = shift;
  my $from;
  if ( ref $options ne 'HASH' ) {
    $from = $options;
    undef $options;
  } else {
    $from = shift;
  }

  my $from_date = DateTime->today( time_zone => 'local' );

  if ( defined $from ) {
    die '"From" argument [ ' . $from . ' ] must be a number'
      unless looks_like_number($from);

    $from_date->subtract( days => $from );
  }

  my @entries = App::Jiffy::TimeEntry::search(
    $self->cfg,
    query => {
      start_time => { '$gt' => $from_date, },
      $self->remove_terminators,
    },
    sort => {
      start_time => 1,
    },
  );

  if ( $options->{round} ) {
    @entries = map { $_->duration( round( $_->duration ) ); $_ } @entries;
  }

  if ( $options->{json} ) {
    my $json = JSON::MaybeXS->new( pretty => 1, convert_blessed => 1 );
    print $json->encode( \@entries );
  } else {
    $options->{from} = $from;
    App::Jiffy::View::Timesheet::render( \@entries, $options );
  }
}

sub search {
  my $self       = shift;
  my $query_text = shift;
  my $options    = shift;
  my $days;
  if ( ref $options ne 'HASH' ) {
    $days = $options;
    undef $options;
  } else {
    $days = shift;
  }

  my $from_date = DateTime->today( time_zone => 'local' );

  if ( defined $days ) {
    die '"Days" argument [ ' . $days . ' ] must be a number'
      unless looks_like_number($days);

    $from_date->subtract( days => $days );
  }

  my @entries = App::Jiffy::TimeEntry::search(
    $self->cfg,
    query => {
      start_time => { '$gt' => $from_date, },
      $self->remove_terminators,
      title => qr/$query_text/,
    },
    sort => {
      start_time => 1,
    },
  );

  if ( $options->{round} ) {
    @entries = map { $_->duration( round( $_->duration ) ); $_ } @entries;
  }

  if ( not @entries and not $options->{json}) {
    print "No Entries Found\n";
    return;
  }

  if ( $options->{json} ) {
    my $json = JSON::MaybeXS->new( pretty => 1, convert_blessed => 1 );
    print $json->encode( \@entries );
  } else {
    $options->{from} = $days;
    App::Jiffy::View::Timesheet::render( \@entries, $options );
  }
}

sub run {
  my $self = shift;
  my @args = @_;

  if ( $args[0] ) {
    if ( $args[0] eq 'current' ) {
      shift @args;
      return $self->current_time(@args);
    } elsif ( $args[0] eq 'timesheet' ) {
      shift @args;

      my $p = Getopt::Long::Parser->new( config => ['pass_through'], );
      $p->getoptionsfromarray(
        \@args,
        'verbose' => \my $verbose,
        'round'   => \my $round,
        'json'    => \my $json
      );

      return $self->time_sheet( {
          verbose => $verbose,
          round   => $round,
          json    => $json,
        },
        @args
      );
    } elsif ( $args[0] eq 'search' ) {
      shift @args;

      my $p = Getopt::Long::Parser->new( config => ['pass_through'], );
      $p->getoptionsfromarray(
        \@args,
        'verbose' => \my $verbose,
        'round'   => \my $round,
        'json'    => \my $json
      );

      my $query_text = shift @args;

      return $self->search(
        $query_text,
        {
          verbose => $verbose,
          round   => $round,
          json    => $json,
        },
        @args
      );
    }
  }

  my $p = Getopt::Long::Parser->new( config => ['pass_through'], );
  $p->getoptionsfromarray( \@args, 'time=s' => \my $time, );

  return $self->add_entry( {
      time => $time,
    },
    join ' ',
    @args
  );
}

1;
__END__

=encoding utf-8

=head1 NAME

App::Jiffy - A minimalist time tracking app focused on precision and effortlessness.

=head1 SYNOPSIS

  use App::Jiffy;

  # cmd line tool
  jiffy Solving world hunger
  jiffy Cleaning the plasma manifolds
  jiffy current # Returns the elapsed time for the current task

  # Run server
  jiffyd
  curl -d "title=Meeting with Client X" http://localhost:3000/timeentry

=head1 DESCRIPTION

App::Jiffy's philosophy is that you should have to do as little as possible to track your time. Instead you should focus on working. App::Jiffy also focuses on precision. Many times time tracking results in globbing activities together masking the fact that your 5 hours of work on project "X" was actually 3 hours of work with interruptions from your coworker asking about project "Y".

In order to be precise with as little effort as possible, App::Jiffy will be available via a myriad of mediums and devices but will have a central server to combine all the information. Plans currently include the following applications:

=over

=item Command line tool

=item Web app L<App::Jiffyd>

=item iPhone app ( potentially )

=back

=head1 INSTALLATION

  curl -L https://cpanmin.us | perl - git://github.com/lejeunerenard/jiffy

=head1 METHODS

The following are methods available on the C<App::Jiffy> object.

=head2 add_entry

C<add_entry> will create a new TimeEntry with the current time as the entry's start_time.

=head2 current_time

C<current_time> will print out the elapsed time for the current task (AKA the time since the last entry was created).

=head2 time_sheet

C<time_sheet> will print out a time sheet including the time spent for each C<TimeEntry>.

=head2 search( C<$query_text>, C<$days> )

The C<search> subcommand will look for the given C<$query_text> in the past C<$days> number of days. It will treat the C<$query_text> argument as a regex.

=head2 run

C<run> will start an instance of the Jiffy app.

=head1 AUTHOR

Sean Zellmer E<lt>sean@lejeunerenard.comE<gt>

=head1 COPYRIGHT

Copyright 2015- Sean Zellmer

=head1 LICENSE

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=cut
